hpx/parallel/algorithms/reduce_by_key.hpp

```diff
@@ -0,0 +1,501 @@
+            }
+        };
+
+        // -------------------------------------------------------------------
+        // helper that extracts the final output iterators from copy_if
+        // -------------------------------------------------------------------
+        // Zip iterator has 3 iterators inside
+        // Iter1, key type : Iter2, value type : Iter3, state type
+        template<typename ZIter, typename iKey, typename iVal>
+        std::pair<iKey, iVal> make_pair_result(ZIter zipiter, iKey key_start,
+            iVal val_start)
+        {
+            // the iterator we want is 'second' part of tagged_pair type (from copy_if)
+            auto const &t = zipiter.second.get_iterator_tuple();
+            iKey key_end = hpx::util::get<0>(t);
+            return std::make_pair(key_end,
+                std::next(val_start, std::distance(key_start, key_end)));
+        }
+
+        // async version that returns future<pair>
+        template<typename ZIter, typename iKey, typename iVal>
+        hpx::future<std::pair<iKey, iVal> > make_pair_result(
+            hpx::future<ZIter> &&ziter, iKey key_start, iVal val_start)
+        {
+            typedef std::pair<iKey, iVal> result_type;
+
+            return lcos::make_future<result_type>(std::move(ziter),
+                [=](ZIter zipiter)
+                {
+                    auto const &t = zipiter.second.get_iterator_tuple();
+                    iKey key_end = hpx::util::get<0>(t);
+                    return std::make_pair(key_end,
+                        std::next(val_start, std::distance(key_start, key_end)));
+                });
+        }
+
+        // -------------------------------------------------------------------
+        // when we are being run with an asynchronous policy, we do not want to
+        // pass the policy directly to other algorithms we are using - as we
+        // would have wait internally on them before proceeding.
+        // Instead create a new policy from the old one which removes the async/future
+        // -------------------------------------------------------------------
+        template<typename ExPolicy>
+        struct remove_asynchronous
+        {
+            typedef ExPolicy type;
+        };
+
+        template<>
+        struct remove_asynchronous<
+            hpx::parallel::parallel_vector_execution_policy
+        >
+        {
+            typedef hpx::parallel::parallel_execution_policy type;
+        };
+
+        template<>
+        struct remove_asynchronous<
+            hpx::parallel::sequential_task_execution_policy
+        >
+        {
+            typedef hpx::parallel::sequential_execution_policy type;
+        };
+
+        template<>
+        struct remove_asynchronous<hpx::parallel::parallel_task_execution_policy>
+        {
+            typedef hpx::parallel::parallel_execution_policy type;
+        };
+
+        /// \endcond
+    }
+
+    // for debugging
+    std::ostream &operator<<(std::ostream &os,
```

Comment:
I'd like to see this being hidden behind a pp constant. Also, if this is enabled, it requires `<iostreams>` to be included.